File: db4on/core/src/core/com/db4o/cs/messages/MsgD.cs
namespace com.db4o.cs.messages
{
	/// <summary>Messages with Data for Client/Server Communication</summary>
	public class MsgD : com.db4o.cs.messages.Msg
	{
		internal com.db4o.YapWriter _payLoad;

		internal MsgD() : base()
		{
		}

		internal MsgD(string aName) : base(aName)
		{
		}

		internal override void FakePayLoad(com.db4o.Transaction a_trans)
		{
		}

		public override com.db4o.YapReader GetByteLoad()
		{
			return _payLoad;
		}

		public sealed override com.db4o.YapWriter PayLoad()
		{
			return _payLoad;
		}

		public com.db4o.cs.messages.MsgD GetWriterForLength(com.db4o.Transaction a_trans, 
			int length)
		{
			com.db4o.cs.messages.MsgD message = (com.db4o.cs.messages.MsgD)Clone(a_trans);
			message._payLoad = new com.db4o.YapWriter(a_trans, length + com.db4o.YapConst.MESSAGE_LENGTH
				);
			message.WriteInt(_msgID);
			message.WriteInt(length);
			if (a_trans.ParentTransaction() == null)
			{
				message._payLoad.Append(com.db4o.YapConst.SYSTEM_TRANS);
			}
			else
			{
				message._payLoad.Append(com.db4o.YapConst.USER_TRANS);
			}
			return message;
		}

		public com.db4o.cs.messages.MsgD GetWriter(com.db4o.Transaction a_trans)
		{
			return GetWriterForLength(a_trans, 0);
		}

		public com.db4o.cs.messages.MsgD GetWriterForInts(com.db4o.Transaction a_trans, int[]
			 ints)
		{
			com.db4o.cs.messages.MsgD message = GetWriterForLength(a_trans, com.db4o.YapConst
				.INT_LENGTH * ints.Length);
			for (int i = 0; i < ints.Length; i++)
			{
				message.WriteInt(ints[i]);
			}
			return message;
		}

		public com.db4o.cs.messages.MsgD GetWriterForIntArray(com.db4o.Transaction a_trans
			, int[] ints, int length)
		{
			com.db4o.cs.messages.MsgD message = GetWriterForLength(a_trans, com.db4o.YapConst
				.INT_LENGTH * (length + 1));
			message.WriteInt(length);
			for (int i = 0; i < length; i++)
			{
				message.WriteInt(ints[i]);
			}
			return message;
		}

		public com.db4o.cs.messages.MsgD GetWriterForInt(com.db4o.Transaction a_trans, int
			 id)
		{
			com.db4o.cs.messages.MsgD message = GetWriterForLength(a_trans, com.db4o.YapConst
				.INT_LENGTH);
			message.WriteInt(id);
			return message;
		}

		public com.db4o.cs.messages.MsgD GetWriterForIntString(com.db4o.Transaction a_trans
			, int anInt, string str)
		{
			com.db4o.cs.messages.MsgD message = GetWriterForLength(a_trans, com.db4o.YapConst
				.stringIO.Length(str) + com.db4o.YapConst.INT_LENGTH * 2);
			message.WriteInt(anInt);
			message.WriteString(str);
			return message;
		}

		public com.db4o.cs.messages.MsgD GetWriterForLong(com.db4o.Transaction a_trans, long
			 a_long)
		{
			com.db4o.cs.messages.MsgD message = GetWriterForLength(a_trans, com.db4o.YapConst
				.LONG_LENGTH);
			message.WriteLong(a_long);
			return message;
		}

		public com.db4o.cs.messages.MsgD GetWriterForString(com.db4o.Transaction a_trans, 
			string str)
		{
			com.db4o.cs.messages.MsgD message = GetWriterForLength(a_trans, com.db4o.YapConst
				.stringIO.Length(str) + com.db4o.YapConst.INT_LENGTH);
			message.WriteString(str);
			return message;
		}

		public virtual com.db4o.cs.messages.MsgD GetWriter(com.db4o.YapWriter bytes)
		{
			com.db4o.cs.messages.MsgD message = GetWriterForLength(bytes.GetTransaction(), bytes
				.GetLength());
			message._payLoad.Append(bytes._buffer);
			return message;
		}

		public virtual byte[] ReadBytes()
		{
			return _payLoad.ReadBytes(ReadInt());
		}

		public int ReadInt()
		{
			return _payLoad.ReadInt();
		}

		public long ReadLong()
		{
			return _payLoad.ReadLong();
		}

		internal sealed override com.db4o.cs.messages.Msg ReadPayLoad(com.db4o.Transaction
			 a_trans, com.db4o.foundation.network.YapSocket sock, com.db4o.YapReader reader)
		{
			int length = reader.ReadInt();
			a_trans = CheckParentTransaction(a_trans, reader);
			com.db4o.cs.messages.MsgD command = (com.db4o.cs.messages.MsgD)Clone(a_trans);
			command._payLoad = new com.db4o.YapWriter(a_trans, length);
			command._payLoad.Read(sock);
			return command;
		}

		public string ReadString()
		{
			int length = ReadInt();
			return com.db4o.YapConst.stringIO.Read(_payLoad, length);
		}

		public void WriteBytes(byte[] aBytes)
		{
			WriteInt(aBytes.Length);
			_payLoad.Append(aBytes);
		}

		internal void WriteInt(int aInt)
		{
			_payLoad.WriteInt(aInt);
		}

		public void WriteLong(long l)
		{
			_payLoad.WriteLong(l);
		}

		public void WriteString(string aStr)
		{
			_payLoad.WriteInt(aStr.Length);
			com.db4o.YapConst.stringIO.Write(_payLoad, aStr);
		}
	}
}

File: db4oj/core/src/com/db4o/internal/cs/messages/MsgD.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o.internal.cs.messages;

import com.db4o.foundation.network.*;
import com.db4o.internal.*;

/**
 * Messages with Data for Client/Server Communication
 */
public class MsgD extends Msg{

	StatefulBuffer _payLoad;

	MsgD() {
		super();
	}

	MsgD(String aName) {
		super(aName);
	}

	public ByteArrayBuffer getByteLoad() {
		return _payLoad;
	}

	public final StatefulBuffer payLoad() {
		return _payLoad;
	}
	
	public void payLoad(StatefulBuffer writer) {
		_payLoad = writer;
	}
	
	public final MsgD getWriterForByte(Transaction trans, byte b) {
		MsgD msg = getWriterForLength(trans, 1);
		msg._payLoad.writeByte(b);
		return msg;
	}
	
	public final MsgD getWriterForLength(Transaction trans, int length) {
		MsgD message = (MsgD)publicClone();
		message.setTransaction(trans);
		message._payLoad = new StatefulBuffer(trans, length + Const4.MESSAGE_LENGTH);
		message.writeInt(_msgID);
		message.writeInt(length);
		if(trans.parentTransaction() == null){
		    message._payLoad.writeByte(Const4.SYSTEM_TRANS);
		}else{
		    message._payLoad.writeByte(Const4.USER_TRANS);
		}
		return message;
	}
	
	public final MsgD getWriter(Transaction trans){
		return getWriterForLength(trans, 0);
	}
	
	public final MsgD getWriterForInts(Transaction trans, int[] ints) {
        MsgD message = getWriterForLength(trans, Const4.INT_LENGTH * ints.length);
        for (int i = 0; i < ints.length; i++) {
            message.writeInt(ints[i]);
        }
        return message;
    }
	
    public final MsgD getWriterForIntArray(Transaction a_trans, int[] ints, int length){
		MsgD message = getWriterForLength(a_trans, Const4.INT_LENGTH * (length + 1));
		message.writeInt(length);
		for (int i = 0; i < length; i++) {
			message.writeInt(ints[i]);
		}
		return message;
	}

	public final MsgD getWriterForInt(Transaction a_trans, int id) {
		MsgD message = getWriterForLength(a_trans, Const4.INT_LENGTH);
		message.writeInt(id);
		return message;
	}
	
	public final MsgD getWriterForIntString(Transaction a_trans,int anInt, String str) {
		MsgD message = getWriterForLength(a_trans, Const4.stringIO.length(str) + Const4.INT_LENGTH * 2);
		message.writeInt(anInt);
		message.writeString(str);
		return message;
	}
	
	public final MsgD getWriterForLong(Transaction a_trans, long a_long){
		MsgD message = getWriterForLength(a_trans, Const4.LONG_LENGTH);
		message.writeLong(a_long);
		return message;
	}
	
	public MsgD getWriterForSingleObject(Transaction trans, Object obj) {
		SerializedGraph serialized = Serializer.marshall(trans.container(), obj);
		MsgD msg = getWriterForLength(trans, serialized.marshalledLength());
		serialized.write(msg._payLoad);
		return msg;
	}

	public final MsgD getWriterForString(Transaction a_trans, String str) {
		MsgD message = getWriterForLength(a_trans, Const4.stringIO.length(str) + Const4.INT_LENGTH);
		message.writeString(str);
		return message;
	}

	public MsgD getWriter(StatefulBuffer bytes) {
		MsgD message = getWriterForLength(bytes.getTransaction(), bytes.length());
		message._payLoad.append(bytes._buffer);
		return message;
	}
	
	public byte[] readBytes(){
	    return _payLoad.readBytes(readInt());
	}

	public final int readInt() {
		return _payLoad.readInt();
	}
	
	public final long readLong(){
	    return _payLoad.readLong();
	}
	
	public final boolean readBoolean() {
		return _payLoad.readByte() != 0;
	}
	
	public Object readObjectFromPayLoad(){
		return Serializer.unmarshall(stream(),_payLoad);
	}

	final Msg readPayLoad(MessageDispatcher messageDispatcher, Transaction a_trans, Socket4 sock, ByteArrayBuffer reader){
		int length = reader.readInt();
		a_trans = checkParentTransaction(a_trans, reader);
		final MsgD command = (MsgD)publicClone();
		command.setTransaction(a_trans);
		command.setMessageDispatcher(messageDispatcher);
		command._payLoad = readMessageBuffer(a_trans, sock, length);
        return command;
	}

	public final String readString() {
		int length = readInt();
		return Const4.stringIO.read(_payLoad, length);
	}
	
	public Object readSingleObject() {
		return Serializer.unmarshall(stream(), SerializedGraph.read(_payLoad));
	}
	
	public final void writeBytes(byte[] aBytes){
	    writeInt(aBytes.length);
	    _payLoad.append(aBytes);
	}

	public final void writeInt(int aInt) {
		_payLoad.writeInt(aInt);
	}
	
	public final void writeLong(long l){
        _payLoad.writeLong(l);
	}

	public final void writeString(String aStr) {
		_payLoad.writeInt(aStr.length());
		Const4.stringIO.write(_payLoad, aStr);
	}

}
